Ray queries

gpx1000 opened the pull request:
## Description

Render a simple scene showing the basics of ray queries, including dynamic shadows top and bottom level acceleration structure use with traditional vertex/fragment shaders. The sample loads the sponza gltf model, creates a bottom acceleration structure for geometry, then makes instances of the geometry. It then demonstrates traversing TLAS (Top Level Acceleration Structure) and BLAS (Bottom Level Acceleration Structure) from the fragment shader using ray queries to cast shadows from a moving light.

## Tested Environments
Windows 10 / Ubuntu Linux
GeForce 3060 / GeForce 2080 / GeForce 1070
Vulkan version: 1.2.176

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

wflohry committed:
Created base for ray queries

wflohry committed:
Updated shaders

wflohry committed:
Restarting ray query demo

wflohry committed:
Adding uniforms and vertex/index buffers

wflohry committed:
Adding descriptor sets / pipeline

wflohry committed:
Updating render pipeline

wflohry committed:
Render pipeline

wflohry committed:
Working ray queries

gpx1000 committed:
Merge branch 'master-upstream' into feature/ray_queries

gpx1000 committed:
ensure clang-format is run and remove warnings.
Add in details in README.md and add to the samples list in samples/CMakeLists.txt

gpx1000 committed:
fix the missing copyright.

JoseEmilio-ARM commented on samples/README.md:
````diff
@@ -181,6 +181,10 @@ Uses the debug utilities extension to name and group Vulkan objects (command buf
 - 🎓 [Using the debug utilities extension](./extensions/debug_utils/debug_utils_tutorial.md)
 
+### [Basic ray queries](./extensions/ray_queries)<br/>
+**Extensions**: [```VK_KHR_RAY_QUERY_EXTENSION_NAME```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_KHR_ray_query), [```VK_KHR_acceleration_structure```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_KHR_acceleration_structure) <br/>
````
We could use VK_KHR_ray_query here for consistency with other entries

JoseEmilio-ARM commented on samples/extensions/ray_queries/ray_queries.h:
```diff
@@ -0,0 +1,109 @@
+
+	struct AccelerationStructure
+	{
+		VkAccelerationStructureKHR         handle         = nullptr;
```
This causes the Android build to fail, fixed with:
```
--- a/samples/extensions/ray_queries/ray_queries.h
+++ b/samples/extensions/ray_queries/ray_queries.h
@@ -68,7 +68,7 @@ class RayQueries : public ApiVulkanSample

        struct AccelerationStructure
        {
-               VkAccelerationStructureKHR         handle         = nullptr;
+               VkAccelerationStructureKHR         handle         = VK_NULL_HANDLE;
                uint64_t                           device_address = 0;
                std::unique_ptr<vkb::core::Buffer> buffer         = nullptr;
        };
@@ -88,10 +88,10 @@ class RayQueries : public ApiVulkanSample
        void                                             create_top_level_acceleration_structure();
        void                                             create_bottom_level_acceleration_structure();

-       VkPipeline            pipeline{nullptr};
-       VkPipelineLayout      pipeline_layout{nullptr};
-       VkDescriptorSet       descriptor_set{nullptr};
-       VkDescriptorSetLayout descriptor_set_layout{nullptr};
+       VkPipeline            pipeline{VK_NULL_HANDLE};
+       VkPipelineLayout      pipeline_layout{VK_NULL_HANDLE};
+       VkDescriptorSet       descriptor_set{VK_NULL_HANDLE};
+       VkDescriptorSetLayout descriptor_set_layout{VK_NULL_HANDLE};

        void build_command_buffers() override;
        void create_uniforms();
```

JoseEmilio-ARM replied:
does not fail the CI though. There are other build warnings that also seem easy to fix: https://github.com/KhronosGroup/Vulkan-Samples/pull/317/checks?check_run_id=3270603265#step:5:366

JoseEmilio-ARM commented on shaders/ray_queries/ray_shadow.frag:
```diff
@@ -0,0 +1,69 @@
+	// whether an intersection exists, and not necessarily any particular intersection
+	rayQueryInitializeEXT(query, topLevelAS, gl_RayFlagsTerminateOnFirstHitEXT, 0xFF, pos, tmin, direction.xyz, distance);
+	while (rayQueryProceedEXT(query))
+	{
```
Empty loops like this can be very expensive on some compilers, suggest using a single call to `rayQueryProceedEXT` here (no more are needed since we are using `gl_RayFlagsTerminateOnFirstHitEXT`)

JoseEmilio-ARM commented on samples/extensions/ray_queries/ray_queries.cpp:
```diff
@@ -0,0 +1,713 @@
+
+		const size_t sz = buffer.get_size();
+		out.resize(sz / sizeof(T));
+		const bool alreadyMapped = buffer.get_data() != nullptr;
```
style: snake_case for variable names

JoseEmilio-ARM replied:
thanks for addressing all of these so quickly! just noticed `bufferName` above too

JoseEmilio-ARM commented on samples/extensions/ray_queries/ray_queries.h:
```diff
@@ -0,0 +1,109 @@
+	RayQueries();
+	~RayQueries() override;
+	void request_gpu_features(vkb::PhysicalDevice &gpu) override;
+	void prepare_render_context() override;
```
no need to override `prepare_render_context`?

gpx1000 replied:
Ah yes, it does the same thing.  Makes sense.

JoseEmilio-ARM commented on samples/extensions/ray_queries/ray_queries.cpp:
```diff
@@ -0,0 +1,713 @@
+	return vkGetBufferDeviceAddressKHR(device->get_handle(), &buffer_device_address_info);
+}
+
+void RayQueries::create_top_level_acceleration_structure()
```
would it be possible to re-use the helper classes introduced in https://github.com/KhronosGroup/Vulkan-Samples/pull/246 ? (@SaschaWillems would the other sample also need to be updated to use them?)

gpx1000 replied:
I didn't do this one yet, leaving it unresolved until I find time to research/implement using Sascha's helper class.

gpx1000 replied:
It looks like we don't have any code in any sample currently making use of @SaschaWillems helper class.  Sascha, please lemme know if you'd like me to help you make one or if you need this sample to be your guinea pig.

Unless Sascha asks for help, I'd like to respectfully allow Sascha to create the first sample using his class as it looks like he has planned and then in a future PR make use of it once any potential bugs are ironed out.  Just as an effort to not step on his toes ;-).

SaschaWillems replied:
Feel free to use and/or change them as and if you like ;) I was under the impression that the ray tracing reflection sample used them, but that seems to have changed for the reworked sample using buffer device address. So yes, they seem to be unused right now.

gpx1000 replied:
@SaschaWillems and @JoseEmilio-ARM I updated the #313  PR to optionally use the framework helper class.  There's a few bugs in the framework so I'd like to let that PR get merged first and come back in a future PR to do the same treatment to this sample if that's okay with you guys?

JoseEmilio-ARM replied:
sounds good, maybe need to raise an issue to not forget

JoseEmilio-ARM replied:
https://github.com/KhronosGroup/Vulkan-Samples/issues/367

JoseEmilio-ARM commented on samples/extensions/ray_queries/ray_queries.cpp:
```diff
@@ -0,0 +1,713 @@
+	std::vector<VkAccelerationStructureBuildRangeInfoKHR *> acceleration_build_structure_range_infos = {&acceleration_structure_build_range_info};
+
+	// Build the acceleration structure on the device via a one-time command buffer submission
+	// Some implementations may support acceleration structure building on the host (VkPhysicalDeviceAccelerationStructureFeaturesKHR->accelerationStructureHostCommands), but we prefer device builds
```
could we rephrase to `but we show device build here`?

JoseEmilio-ARM commented on shaders/ray_queries/ray_shadow.frag:
```diff
@@ -0,0 +1,69 @@
+	rayQueryEXT query;
+
+	// The following runs the actual ray query
+	// For performance, could use gl_RayFlagsTerminateOnFirstHitEXT, since we only need to know
```
can remove `could`?

gpx1000 committed:
fix warnings
VK_KHR_ray_query in the README.md
fix reported Android build fail
remove empty loop and replace with explanation of why/what gl_RayFlagsTerminateOnFirstHitEXT does
snake_case fix
remove prepare_render_context
rephrase to suggestion at ray_queries.cpp:455
remove could at ray_shadow.frag:50

gpx1000 committed:
fix clang-format

gpx1000 committed:
device build clarification.

gpx1000 committed:
fix camel case to snake case

JoseEmilio-ARM commented on samples/extensions/ray_queries/ray_queries.cpp:
```diff
@@ -0,0 +1,708 @@
+		VkDeviceSize offsets[1] = {0};
+		vkCmdBindVertexBuffers(draw_cmd_buffers[i], 0, 1, vertex_buffer->get(), offsets);
+		vkCmdBindIndexBuffer(draw_cmd_buffers[i], index_buffer->get_handle(), 0, VK_INDEX_TYPE_UINT32);
+		vkCmdDrawIndexed(draw_cmd_buffers[i], static_cast<uint32_t>(model.indices.size()) * 3, 1, 0, 0, 0);
```
It seems that we are aggregating all geometry and dispatching a single draw call for the whole scene. Perhaps a more realistic use case could be to have a deferred rendering pipeline, to use the same draw calls to render both the scene and the shadows, since the ray queries can then be traced once per pixel during the lighting pass

gpx1000 replied:
I'm less certain about this, it's a single model.  A good optimization for most use cases is to reduce the number of draw calls and this is emblematic of that.  However, at the same time it would make sense to use space partitioning given that the model is a full scene one can traverse.  The complexity rises to do this yet doesn't refine the actual sample's use case.
Maybe a good compromise would be to add a model to the scene that explicitly could be a good use case for another draw call?  What are your thoughts?

JoseEmilio-ARM replied:
I was more concerned about overdraw, since we might be unnecessarily running ray queries for occluded pixels. A deferred pass would limit it to one ray per pixel. I can see how this would add complexity to the sample though.

gpx1000 replied:
Overdraw is a compelling reason.  I have to admit to being torn as I think simple samples are better than complex ones to communicate the core idea.  However, a reasoning for ray queries to exist in a common pipeline would be to be able to make use of running them only in a deferred fragment pass so they are limited in scope to the number of pixels actually to be rendered.
I'm leaning towards saying this is beyond the scope of a basic introduction to ray queries as a core tenet of the extension is they can be used from any shader at any point in either the traditional or ray tracing pipelines.  However, a performance sample would make a lot of sense to demonstrate the reasoning why overdraw would be a potential issue.  Maybe a future sample should address this.  Do you have any thoughts on that approach?

JoseEmilio-ARM replied:
> a reasoning for ray queries to exist in a common pipeline would be to be able to make use of running them only in a deferred fragment pass so they are limited in scope to the number of pixels actually to be rendered

Yes exactly, I was going for a simple deferred pipeline like in some of the performance samples, to show how ray queries can be part of it. But also happy to let this sample focus on the use of the extension and address more realistic use-cases in future samples

JoseEmilio-ARM commented on shaders/ray_queries/ray_shadow.frag:
```diff
@@ -0,0 +1,72 @@
+	// For performance, use gl_RayFlagsTerminateOnFirstHitEXT, since we only need to know
+	// whether an intersection exists, and not necessarily any particular intersection
+	rayQueryInitializeEXT(query, topLevelAS, gl_RayFlagsTerminateOnFirstHitEXT, 0xFF, pos, tmin, direction.xyz, distance);
+	// The following is the cononical way of using ray Queries from the fragment shader when
```
`cononical` -> `canonical`

gpx1000 committed:
typo fix

SaschaWillems commented on samples/extensions/ray_queries/ray_queries.h:
```diff
@@ -0,0 +1,108 @@
+	void draw();
+
+	uint32_t max_thread_count{1};
+	bool     enable_shadows{false};
```
These two variables are not used in the sample. Maybe some leftovers?

SaschaWillems commented:
Just gave this a spin and I'm happy with this. 

I have one suggestion though: Adding at least basic lighting to the scene. With the scene only being a single-colored mesh, it kinda looks odd if you're looking at a mostly shadowed area, with just two different grey values.

And will there be a markdown tutorial for this sample too?

wflohry committed:
Removed unused variables

gpx1000 committed:
Merge branch 'feature/ray_queries' into ray_queries

wflohry committed:
Added AO to improve ray queries sample

gpx1000 committed:
Merge branch 'feature/ray_queries' into ray_queries

gpx1000 committed:
run clang-format for fix.

gpx1000 commented:
> Just gave this a spin and I'm happy with this.
> 
> I have one suggestion though: Adding at least basic lighting to the scene. With the scene only being a single-colored mesh, it kinda looks odd if you're looking at a mostly shadowed area, with just two different grey values.
> 
> And will there be a markdown tutorial for this sample too?

Thanks @SaschaWillems for the review
We added an AO per your request.

It's my hope that the code in the sample is a sufficient tutorial in this case as to where a markdown tutorial would be superfluous.  If we get a lot of questions about it, it'd make sense to add one in a future PR to make certain everything is explicit.  Otherwise, I'm inclined to recommend the code is simple enough to not benefit from an explicit tutorial.

SaschaWillems commented:
> Thanks @SaschaWillems for the review
We added an AO per your request.

Thanks. That's much better :) One minor suggestion would be lowering the AO range a bit, as the effect looks a bit odd at first.

> It's my hope that the code in the sample is a sufficient tutorial in this case as to where a markdown tutorial would be superfluous. If we get a lot of questions about it, it'd make sense to add one in a future PR to make certain everything is explicit. Otherwise, I'm inclined to recommend the code is simple enough to not benefit from an explicit tutorial.

Sure, I'm fine with that too :)

JoseEmilio-ARM commented on shaders/ray_queries/ray_shadow.frag:
```diff
@@ -0,0 +1,121 @@
+
+			rayQueryEXT query;
+			rayQueryInitializeEXT(query, topLevelAS, gl_RayFlagsTerminateOnFirstHitEXT, 0xFF, object_point, tmin, direction.xyz, tmax);
+			while (rayQueryProceedEXT(query))
```
similar to `intersects_light` below, please avoid loop

gpx1000 replied:
Do you have a suggestion for what to change it to?
I was under the impression that rayQueryProceedExt is meant to return false until it is finished traversing.  gl_RayFlagsTerminateOnFirstHitExp should hopefully ensure this is only called once.  If it's changed to a different flag, then the while loop might have a real cost but doesn't this combo result in a no cost loop that shows correct usage?

JoseEmilio-ARM replied:
ah sorry, I meant to replace it with exactly the same as in lines 100-105 below, to call `rayQueryProceed` by itself rather than in an empty while loop

gpx1000 replied:
Sorry still a little confused, line 100 has an empty while loop as well.

Do you mean the nested for loop logic?  Move the while loop outside the for loop(s)?  This should run 9 times as written with a no-op essentially on the while loop as it's going to be a first hit.

JoseEmilio-ARM replied:
Hi, apologies for the delay. I mean the empty `while` loop, just for consistency and also because in this case a single call is enough and avoids compiler costs, as explained in the comments below in lines 100-105. Might not need to repeat the comments in both places though

```
	// The following is the canonical way of using ray Queries from the fragment shader when
	// there's more than one bounce or hit to traverse:
	// while (rayQueryProceedEXT(query)) { }
	// This sample has set flags to gl_RayFlagsTerminateOnFirstHitEXT which means that there
	// will never be a bounce and no need for an expensive while loop.  (i.e. we only need to call it once).
	rayQueryProceedEXT(query);
```

Happy to approve after this is resolved 🙏

gpx1000 committed:
Merge branch 'master-upstream' into ray_queries

gpx1000 committed:
remove while loop to explicitly show that RayQueryProceedEXT is only run once.  NB: The correct thing to do when running RayQueryProceedEXT when there's more than one bounce is to use a while loop.  As this sample has no extra calls by design (gl_RayFlagsTerminatedOnFirstHitEXT is used and while loop is dropped.

SaschaWillems reviewed (approved):
LGTM. Runs fine on my RTX 2060 :)
